File: confirmation-email-sent-before-payment/python/flag_confirmation_timing.py
"""Flag Saleor orders whose confirmation email fired before the payment ever
succeeded, because send_order_confirmation runs synchronously inside
checkoutComplete at order-creation time, never gated on a successful
CHARGE_SUCCESS or AUTHORIZATION_SUCCESS transaction event (see saleor/saleor#3527
and the TransactionEvent and Order object docs).

This script never un-sends an email, since Saleor has no mutation for that.
Under DRY_RUN=true (the default) it only logs a report entry for each flagged
order for support follow-up. Only orders that clear CANCEL_GRACE_HOURS with
still no successful charge are cancelled with orderCancel, and only when
DRY_RUN=false. Run on a schedule. Safe to run again and again.
"""
import os
import logging
import datetime
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("flag_confirmation_timing")

API_URL = os.environ.get("SALEOR_API_URL", "https://store.saleor.cloud/graphql/")
TOKEN = os.environ.get("SALEOR_AUTH_TOKEN", "dummy-token")
CANCEL_GRACE_HOURS = float(os.environ.get("CANCEL_GRACE_HOURS", "24"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

ORDERS_QUERY = """
query OrdersWithTimeline($first: Int!, $after: String) {
  orders(first: $first, after: $after) {
    pageInfo { hasNextPage endCursor }
    edges {
      node {
        id number isPaid status created
        events { type date }
        payments { id isActive transactions { id kind isSuccess created } }
      }
    }
  }
}"""

TRANSACTIONS_QUERY = """
query OrderTransactions($id: ID!) {
  order(id: $id) {
    id isPaid
    transactions { id events { type createdAt pspReference } }
  }
}"""

CANCEL_ORDER = """
mutation CancelUnpaidOrder($id: ID!) {
  orderCancel(id: $id) {
    order { id status }
    errors { field message code }
  }
}"""


def gql(query, variables=None):
    r = requests.post(
        API_URL,
        json={"query": query, "variables": variables or {}},
        headers={"Authorization": f"Bearer {TOKEN}", "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def confirm_event_timestamp(order):
    """Return the ISO timestamp of the order's PLACED event, or None."""
    for event in order.get("events") or []:
        if event.get("type") == "PLACED":
            return event.get("date")
    return None


def charge_success_timestamp(order, transactions):
    """Return the earliest successful charge timestamp for the order, or None.

    Prefers the current Transactions API (TransactionEvent type CHARGE_SUCCESS)
    and falls back to the legacy Payment/Transaction model (kind CAPTURE,
    isSuccess true) for deployments still on the old payments flow.
    """
    times = [
        e["createdAt"]
        for t in (transactions or [])
        for e in (t.get("events") or [])
        if e.get("type") == "CHARGE_SUCCESS"
    ]
    if not times:
        for payment in order.get("payments") or []:
            for tx in payment.get("transactions") or []:
                if tx.get("kind") == "CAPTURE" and tx.get("isSuccess"):
                    times.append(tx["created"])
    return min(times) if times else None


def decide_confirmation_timing_issue(confirm_event_ts, charge_success_ts, order_is_paid,
                                      now, cancel_grace_hours=24):
    """Pure decision function. No I/O, fully unit-testable.

    confirm_event_ts: datetime | None - when the PLACED event (and therefore
        send_order_confirmation) fired.
    charge_success_ts: datetime | None - earliest successful charge event, if any.
    order_is_paid: bool - order.isPaid at flag time.
    now: datetime - current time, passed in so the function stays pure.
    cancel_grace_hours: int - how long to wait before an unpaid order with a
        premature confirmation becomes eligible for cancellation.

    Returns one of "ok", "flag_email_premature", "flag_and_eligible_for_cancel".
    """
    if confirm_event_ts is None:
        return "ok"
    if charge_success_ts is not None and confirm_event_ts <= charge_success_ts:
        return "ok"
    if charge_success_ts is not None and confirm_event_ts > charge_success_ts:
        return "ok"
    if order_is_paid:
        return "ok"
    age_hours = (now - confirm_event_ts).total_seconds() / 3600
    if age_hours >= cancel_grace_hours:
        return "flag_and_eligible_for_cancel"
    return "flag_email_premature"


def parse_iso(value):
    return datetime.datetime.fromisoformat(value.replace("Z", "+00:00"))


def all_orders():
    cursor = None
    while True:
        data = gql(ORDERS_QUERY, {"first": 50, "after": cursor})["orders"]
        for edge in data["edges"]:
            yield edge["node"]
        if not data["pageInfo"]["hasNextPage"]:
            return
        cursor = data["pageInfo"]["endCursor"]


def fetch_transactions(order_id):
    return gql(TRANSACTIONS_QUERY, {"id": order_id})["order"]["transactions"]


def cancel_unpaid_order(order_id):
    result = gql(CANCEL_ORDER, {"id": order_id})["orderCancel"]
    if result["errors"]:
        raise RuntimeError(result["errors"])
    return result["order"]["status"]


def run():
    now = datetime.datetime.now(datetime.timezone.utc)
    flagged = 0
    cancelled = 0

    for order in all_orders():
        confirm_iso = confirm_event_timestamp(order)
        if confirm_iso is None:
            continue

        transactions = fetch_transactions(order["id"])
        charge_iso = charge_success_timestamp(order, transactions)

        confirm_ts = parse_iso(confirm_iso)
        charge_ts = parse_iso(charge_iso) if charge_iso else None

        outcome = decide_confirmation_timing_issue(
            confirm_ts, charge_ts, order.get("isPaid", False), now, CANCEL_GRACE_HOURS
        )
        if outcome == "ok":
            continue

        flagged += 1
        report_entry = {
            "orderId": order["id"],
            "number": order["number"],
            "confirmEventTs": confirm_iso,
            "chargeSuccessTs": charge_iso,
            "isPaid": order.get("isPaid", False),
            "status": order.get("status"),
            "outcome": outcome,
        }
        log.warning("Confirmation timing issue found. %s", report_entry)

        if outcome == "flag_and_eligible_for_cancel":
            if not DRY_RUN:
                cancel_unpaid_order(order["id"])
                cancelled += 1
            else:
                log.info("Order %s would be cancelled (dry run).", order["number"])

    log.info("Done. %d order(s) flagged, %d cancelled.", flagged, cancelled)


if __name__ == "__main__":
    run()

File: confirmation-email-sent-before-payment/node/flag-confirmation-timing.js
/**
 * Flag Saleor orders whose confirmation email fired before the payment ever
 * succeeded, because send_order_confirmation runs synchronously inside
 * checkoutComplete at order-creation time, never gated on a successful
 * CHARGE_SUCCESS or AUTHORIZATION_SUCCESS transaction event (see saleor/saleor#3527
 * and the TransactionEvent and Order object docs).
 *
 * This script never un-sends an email, since Saleor has no mutation for that.
 * Under DRY_RUN=true (the default) it only logs a report entry for each flagged
 * order for support follow-up. Only orders that clear CANCEL_GRACE_HOURS with
 * still no successful charge are cancelled with orderCancel, and only when
 * DRY_RUN=false. Run on a schedule.
 *
 * Guide: https://www.allanninal.dev/saleor/confirmation-email-sent-before-payment/
 */
import { pathToFileURL } from "node:url";

const API_URL = process.env.SALEOR_API_URL || "https://store.saleor.cloud/graphql/";
const TOKEN = process.env.SALEOR_AUTH_TOKEN || "dummy-token";
const CANCEL_GRACE_HOURS = Number(process.env.CANCEL_GRACE_HOURS || 24);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

/** Return the ISO timestamp of the order's PLACED event, or null. */
export function confirmEventTimestamp(order) {
  for (const event of order.events || []) {
    if (event.type === "PLACED") return event.date;
  }
  return null;
}

/**
 * Return the earliest successful charge timestamp for the order, or null.
 * Prefers the current Transactions API (TransactionEvent type CHARGE_SUCCESS)
 * and falls back to the legacy Payment/Transaction model (kind CAPTURE,
 * isSuccess true) for deployments still on the old payments flow.
 */
export function chargeSuccessTimestamp(order, transactions) {
  let times = [];
  for (const t of transactions || []) {
    for (const e of t.events || []) {
      if (e.type === "CHARGE_SUCCESS") times.push(e.createdAt);
    }
  }
  if (times.length === 0) {
    for (const payment of order.payments || []) {
      for (const tx of payment.transactions || []) {
        if (tx.kind === "CAPTURE" && tx.isSuccess) times.push(tx.created);
      }
    }
  }
  return times.length ? times.sort()[0] : null;
}

/**
 * Pure decision function. No I/O, fully unit-testable.
 *
 * confirmEventTs: epoch millis | null - when the PLACED event (and therefore
 *   send_order_confirmation) fired.
 * chargeSuccessTs: epoch millis | null - earliest successful charge event, if any.
 * orderIsPaid: boolean - order.isPaid at flag time.
 * now: epoch millis - current time, passed in so the function stays pure.
 * cancelGraceHours: number - how long to wait before an unpaid order with a
 *   premature confirmation becomes eligible for cancellation.
 *
 * Returns one of "ok", "flag_email_premature", "flag_and_eligible_for_cancel".
 */
export function decideConfirmationTimingIssue(confirmEventTs, chargeSuccessTs, orderIsPaid,
                                               now, cancelGraceHours = 24) {
  if (confirmEventTs === null || confirmEventTs === undefined) return "ok";
  if (chargeSuccessTs !== null && chargeSuccessTs !== undefined && confirmEventTs <= chargeSuccessTs) return "ok";
  if (chargeSuccessTs !== null && chargeSuccessTs !== undefined && confirmEventTs > chargeSuccessTs) return "ok";
  if (orderIsPaid) return "ok";
  const ageHours = (now - confirmEventTs) / (1000 * 60 * 60);
  if (ageHours >= cancelGraceHours) return "flag_and_eligible_for_cancel";
  return "flag_email_premature";
}

async function gql(query, variables = {}) {
  const res = await fetch(API_URL, {
    method: "POST",
    headers: { Authorization: `Bearer ${TOKEN}`, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Saleor ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const ORDERS_QUERY = `
query OrdersWithTimeline($first: Int!, $after: String) {
  orders(first: $first, after: $after) {
    pageInfo { hasNextPage endCursor }
    edges {
      node {
        id number isPaid status created
        events { type date }
        payments { id isActive transactions { id kind isSuccess created } }
      }
    }
  }
}`;

const TRANSACTIONS_QUERY = `
query OrderTransactions($id: ID!) {
  order(id: $id) {
    id isPaid
    transactions { id events { type createdAt pspReference } }
  }
}`;

const CANCEL_ORDER = `
mutation CancelUnpaidOrder($id: ID!) {
  orderCancel(id: $id) {
    order { id status }
    errors { field message code }
  }
}`;

async function* allOrders() {
  let cursor = null;
  while (true) {
    const data = (await gql(ORDERS_QUERY, { first: 50, after: cursor })).orders;
    for (const edge of data.edges) yield edge.node;
    if (!data.pageInfo.hasNextPage) return;
    cursor = data.pageInfo.endCursor;
  }
}

async function fetchTransactions(orderId) {
  const data = await gql(TRANSACTIONS_QUERY, { id: orderId });
  return data.order.transactions;
}

async function cancelUnpaidOrder(orderId) {
  const result = (await gql(CANCEL_ORDER, { id: orderId })).orderCancel;
  if (result.errors.length) throw new Error(JSON.stringify(result.errors));
  return result.order.status;
}

export async function run() {
  const now = Date.now();
  let flagged = 0;
  let cancelled = 0;

  for await (const order of allOrders()) {
    const confirmIso = confirmEventTimestamp(order);
    if (confirmIso === null) continue;

    const transactions = await fetchTransactions(order.id);
    const chargeIso = chargeSuccessTimestamp(order, transactions);

    const confirmTs = Date.parse(confirmIso);
    const chargeTs = chargeIso ? Date.parse(chargeIso) : null;

    const outcome = decideConfirmationTimingIssue(
      confirmTs, chargeTs, order.isPaid === true, now, CANCEL_GRACE_HOURS
    );
    if (outcome === "ok") continue;

    flagged++;
    const reportEntry = {
      orderId: order.id,
      number: order.number,
      confirmEventTs: confirmIso,
      chargeSuccessTs: chargeIso,
      isPaid: order.isPaid === true,
      status: order.status,
      outcome,
    };
    console.warn("Confirmation timing issue found.", reportEntry);

    if (outcome === "flag_and_eligible_for_cancel") {
      if (!DRY_RUN) {
        await cancelUnpaidOrder(order.id);
        cancelled++;
      } else {
        console.log(`Order ${order.number} would be cancelled (dry run).`);
      }
    }
  }

  console.log(`Done. ${flagged} order(s) flagged, ${cancelled} cancelled.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
